src/history/hs_verify.c

```diff
@@ -0,0 +1,230 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __hs_verify_id --
+ *     Verify the history store for a single btree. Given a cursor to the tree, walk all history
+ *     store keys. This function assumes any caller has already opened a cursor to the history
+ *     store.
+ */
+static int
+__hs_verify_id(
+  WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, WT_CURSOR_BTREE *ds_cbt, uint32_t this_btree_id)
+{
+    WT_DECL_ITEM(prev_key);
+    WT_DECL_RET;
+    WT_ITEM key;
+    wt_timestamp_t hs_start_ts;
+    uint64_t hs_counter, recno;
+    uint32_t btree_id;
+    const uint8_t *up;
+    int cmp;
+
+    WT_CLEAR(key);
+    WT_ERR(__wt_scr_alloc(session, 0, &prev_key));
+
+    /*
+     * If using standard cursors, we need to skip the non-globally visible tombstones in the data
+     * table to verify the corresponding entries in the history store are too present in the data
+     * store. Though this is not required currently as we are directly searching btree cursors,
+     * leave it here in case we switch to standard cursors.
+     */
+    F_SET(&ds_cbt->iface, WT_CURSTD_IGNORE_TOMBSTONE);
+
+    /*
+     * The caller is responsible for positioning the history store cursor at the first record to
+     * verify. When we return after moving to a new key the caller is responsible for keeping the
+     * cursor there or deciding they're done.
+     */
+    for (; ret == 0; ret = hs_cursor->next(hs_cursor)) {
+        /*
+         * If the btree id does not match the previous one, we're done. It is up to the caller to
+         * set up for the next tree and call us, if they choose.
+         */
+        WT_ERR(hs_cursor->get_key(hs_cursor, &btree_id, &key, &hs_start_ts, &hs_counter));
+        if (btree_id != this_btree_id)
+            break;
+
+        /*
+         * If we have already checked against this key, keep going to the next key. We only need to
+         * check the key once.
+         */
+        WT_ERR(__wt_compare(session, NULL, &key, prev_key, &cmp));
```

Comment:
I was thinking rather than using the `prev_key` and checking if they are the same key, to just check if the `hs_counter` was incremented meaning that there is a duplicate key

Reply:
I don't think we can rely on the `hs_counter` straight away.

Before comparing the `hs_counter`, we have to make sure the `btree_id`, the `key` and the `hs_start_ts` are the same. Indeed, only when those fields are identical, we increment the `hs_counter` by 1.

@kommiharibabu, please correct me if I am wrong.